File: TestLeetCode/Backtrack/SubarraySum.cs
// https://leetcode.com/problems/combination-sum/description/
// https://www.youtube.com/watch?v=GBKI9VSKdGg

public class CombinationSumSolution {
    IList<IList<int>> result = new List<IList<int>>();
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        void backtrack(int index, List<int> path, int total) {
            if(total == target) {
                result.Add(path.ToList());
                return;
            }
        
            if(total > target || index >= candidates.Length) return;
            
            path.Add(candidates[index]);
            backtrack(index, 
                    path, 
                    total + candidates[index]
                    );
            
            path.Remove(path.Last());
            
            backtrack(index + 1, 
                    path, 
                    total
                    );
        }

        backtrack(0, new List<int>(), 0);
        return result;
    }
}

File: python/backtrack/subarray_sum.py
class CombinationSum:
    def combination_sum(self, candidates: list[int], target: int) -> list[list[int]]:
        result = []
        def visit(index: int, total: int, path: list[int]) -> None:
            if total == target:
                result.append(path.copy()); return
            if total > target or index == len(candidates):
                return
            path.append(candidates[index]); visit(index, total + candidates[index], path); path.pop()
            visit(index + 1, total, path)
        visit(0, 0, [])
        return result
